source/extensions/filters/http/oauth/oauth_client.cc

```diff
@@ -0,0 +1,102 @@
+#include "extensions/filters/http/oauth/oauth_client.h"
+
+#include <chrono>
+
+#include "envoy/http/async_client.h"
+#include "envoy/http/message.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "common/common/fmt.h"
+#include "common/common/logger.h"
+#include "common/http/message_impl.h"
+#include "common/protobuf/message_validator_impl.h"
+
+#include "source/extensions/filters/http/oauth/oauth_response.pb.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Oauth {
+
+namespace {
+Http::RegisterCustomInlineHeader<Http::CustomInlineHeaderRegistry::Type::RequestHeaders>
+    authorization_handle(Http::CustomHeaders::get().Authorization);
+
+constexpr const char* GetAccessTokenBodyFormatString =
+    "grant_type=authorization_code&code={0}&client_id={1}&client_secret={2}&redirect_uri={3}";
+
+} // namespace
+
+void OAuth2ClientImpl::asyncGetAccessToken(const std::string& auth_code,
+                                           const std::string& client_id, const std::string& secret,
+                                           const std::string& cb_url) {
+  Http::RequestMessagePtr request = createPostRequest();
+  request->headers().setPath(oauth_token_path_);
+  const std::string body =
+      fmt::format(GetAccessTokenBodyFormatString, auth_code, client_id, secret, cb_url);
```

Comment:
I'm not an OAuth expert, but just would like to check whether any of these substitutions might need some form of escape encoding?

Reply:
cc: @jparise

Reply:
Yes, because this is being sent as `application/x-www-form-urlencoded` (`ContentTypeValues.FormUrlEncoded)`, the POST body values should be form-encoded.

Reply:
I guess this becomes a little bit tricky to manage since we want to avoid double URL encoding these parameters. How about this: 

auth_code is read in from a query param, so should already be safe
client_id/secret are configured opaque strings - these can be documented as having to be URL encoded and validated as such during config load. 
cb_url can be percent encoded in code so that users can configure a human readable url

I think this would nicely balance API usability and the risk that we double encode anything

Reply:
Alternatively we could always percent encode client_id/secret/cb_url, I guess either way works. Thoughts on this from an API perspective?

Reply:
@snowp I think your reasoning makes sense, and it seems safe (and obvious) to always percent-encode those last three values.